util/rom-ext-manifest-generator.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import copy
+from pathlib import Path
+
+import hjson
+from mako.template import Template
+
+from topgen.c import MemoryRegion, Name
+
+DESC = """ROM_EXT manifest generator"""
+
+USAGE = """
+  rom-ext-manifest-generator --input-dir:
+    Directory where manifest.hjson and manifest.h.tpl reside.
+
+  rom-ext-manifest-generator --output-dir:
+    Directory where manifest.hjson and manifest.h.tpl reside.
+"""
+
+
+def add_size_in_bytes(fields):
+    """Adds `size_in_bytes` entry to every parsed ROM_EXT manifest field.
+
+    fields[N]    =>    fields[N]
+      - ...              - ...
+      - size             - size
+                         - size_in_bytes
+    """
+
+    for field in fields:
+        field['size_in_bytes'] = field['size'] // 8
+
+
+def add_offsets_in_bytes(fields):
+    """Adds `offset_in_bytes` entry to every parsed ROM_EXT manifest field.
+
+    fields[N]    =>    fields[N]
+      - ...              - ...
+      - size_in_bytes    - size_in_bytes
+                         - offset_in_bytes
+    """
+
+    accumulated_offset = 0
+
+    for field in fields:
+        field['offset'] = hex(accumulated_offset)
+        accumulated_offset += field['size_in_bytes']
+
+
+def generate_cheader(fields, input_dir, output_dir):
+    """ Generates C header file from the `template_file`.
+
+    It produces a list of tuples with a field name and the `MemoryRegion`
+    object, whic is used in the `template_file`. The resulting header file is
+    placed into `output_dir`.
+    """
+
+    manifest_cheader_template_file = input_dir / 'manifest.h.tpl'
+    manifest_cheader_output_file = output_dir / 'manifest.h'
+
+    base_name = Name.from_snake_case("ROM_EXT")
+
+    items = []
+    for field in fields:
+        if field['type'] == "field":
+            item = (field['name'],
+                    MemoryRegion(
+                        base_name + Name.from_snake_case(field['name']),
+                        field['offset'], field['size_in_bytes']))
```

Comment:
This would take fewer lines and be easier to read, with temporary variables. I'd suggest something like this:
```python
    items = []
    for field in fields:
        if field['type'] == "field":
            region_name = base_name + Name.from_snake_case(field['name'])
            region = MemoryRegion(region_name,
                                  field['offset'], field['size_in_bytes'])
            items.append((field['name'], region))
```

Of course, since the first thing that `generate_cheader` does is to iterate over the fields, it would probably be cleaner to calculate "offset" and "size_in_bytes" on the fly (getting rid of `add_size_in_bytes`, `add_offsets_in_bytes` and all the copying). Something like this:

```python
    items = []
    offset = 0
    for field in fields:
        assert field['size'] % 8 == 0
        size_bytes = field['size'] // 8
        if field['type'] == "field":
            region_name = base_name + Name.from_snake_case(field['name'])
            region = MemoryRegion(region_name, hex(offset), size_bytes)
            items.append((field['name'], region))

        offset += size_bytes
```

This is barely longer than the loop you had before and gets rid of all the helper functions.

Reply:
Thank you @rswarbrick !I will implement all your suggestions `-` the offset helper (personal preference - as this field is fundamental and would be used by other generators like `.S` if we wish to use it). I agree, at this point we have only C header generation, but I would like to keep the offsets helper anyway.

Reply:
On the other hand, I will just go ahead and implement your suggestion entirely, and if we need to we can come back to it later.

Reply:
Done.